data_prep/project_context/context_introspector.py

```diff
@@ -0,0 +1,197 @@
+    # query the definition for.
+    for curr_line in range(begin_line, begin_line_elem):
+      source_line = introspector.query_introspector_source_code(
+          self._benchmark.project, file_name, curr_line, curr_line)
+      reconstructed_type += source_line
+
+    for curr_line in range(begin_line_elem, end_line_elem + 1):
+      source_line = introspector.query_introspector_source_code(
+          self._benchmark.project, file_name, curr_line, curr_line)
+      reconstructed_type += source_line
+
+      newly_seen_type = self._clean_type(
+          self._extract_type_from_source_line(
+              source_line, elements[curr_line - begin_line_elem]))
+
+      if newly_seen_type and newly_seen_type not in seen_types:
+        seen_types.add(newly_seen_type)
+        types_to_get.add(newly_seen_type)
+
+    # If we do not see a '}' in the final element's source line
+    # Then we can add it ourselves. This would cause problems
+    # when typedef and struct definitions are combined.
+    # The alternative is to query for source code lines until a '}' is found.
+    if '}' not in source_line:
+      reconstructed_type += '};\n'
+
+    return reconstructed_type
+
+  def get_embeddable_types(self) -> list[str]:
+    """Retrieves types from FI."""
+    seen_types = set()
+    types_to_get = set()
+    types = set()
+
+    types_to_get.add(self._clean_type(self._benchmark.return_type))
+
+    params = self._benchmark.params
+
+    for param in params:
+      cleaned_type = self._clean_type(param['type'])
+      if cleaned_type:
+        types_to_get.add(cleaned_type)
+
+    seen_types = types_to_get.copy()
+
+    # Add support for recursively querying for types
+    iteration = 0
+    while types_to_get:
+      iteration += 1
+      current_type = types_to_get.pop()
+      print(f'Querying for type: {current_type}')
+      info = introspector.query_introspector_type_info(self._benchmark.project,
+                                                       current_type)
+      if not info:
+        continue
+
+      reconstructed_type = ''
+      # For now, info is a single element and not a list.
+      # Requested it be changed to a list, so that type
+      # elements with the same name will be returned.
+      if info['type'] == 'struct':
+        reconstructed_type = self._get_struct_type(info, seen_types,
+                                                   types_to_get)
+      if info['type'] == 'typedef':
+        reconstructed_type = self._get_typedef_type(info)
+
+      if reconstructed_type:
+        types.add(reconstructed_type)
+
+    return list(types)
+
+  def _extract_type_from_source_line(self, source_line: str,
+                                     type_element: dict) -> str:
+    """Attempts to extract a type from a source line.
+    Do so by attempting to match for name and extracting
+    everything before it."""
+    start = source_line.find(next(filter(str.isalpha, source_line)))
```

Comment:
nit: 

```
next(i for i, c in enumerate(source_line) if str.isalpha(c))
```

might be slightly easier to understand here, as it's less of a roundabout way of achieving this.

Reply:
done thanks